### ISSUE TITLE
Training Continuation Causes Autodiff Error For Cuda and Pytorch Backends/Devices

### ISSUE THREAD

**Issue by fuzing** [tracel-ai/burn#5082]
I've got a transformer model that was trained for a couple of epochs using SupervisedTraining (a multi-label classification task).  The original SupervisedTraining instantiation used .num_epochs(2).

After deciding to continue the training for a couple more epochs I instantiated the SupervisedTraining with a call to .checkpoint(2) and increased to .num_epochs(4).

This barfed using Device::libtorch_cuda(DeviceIndex::Default) with an error related to not using an autodiff backend.

To rule out a problem with libtorch I subsequently downgraded to cuda-toolkit v12.5 and went directly with the cuda backend (i.e. Device::cuda(DeviceIndex::Default)).  The error I'm getting now is similar to that when using the libtorch device, namely:

```
thread 'main' (32587) panicked at /home/some-folder/burn/crates/burn-dispatch/src/ops/tensor.rs:54:9

Cannot move between autodiff and non-autodiff instances
```

Boilerplate for the SupervisedTraining setup looks like:
```
let mut training = SupervisedTraining::new(artifact_path, dataloader_train, dataloader_validation)
        .metric_train(CudaMetric::new())
        .metric_valid(CudaMetric::new())
        .metric_train(IterationSpeedMetric::new())
        .metric_train_numeric(LossMetric::new())
        .metric_valid_numeric(LossMetric::new())
        .metric_train_numeric(AccuracyMetric::new())
        .metric_valid_numeric(AccuracyMetric::new())

        .metric_train_numeric(RecallMetric::multiclass(1, ClassReduction::Macro))
        .metric_valid_numeric(RecallMetric::multiclass(1, ClassReduction::Macro))
        .metric_train_numeric(PrecisionMetric::multiclass(1, ClassReduction::Macro))
        .metric_valid_numeric(PrecisionMetric::multiclass(1, ClassReduction::Macro))

        .metric_train_numeric(LearningRateMetric::new())
        .with_file_checkpointer(CompactRecorder::new())
        .with_training_strategy(strategy.into())
        .num_epochs(args.num_epochs)
        .summary();

    // resuming training from particular checkpoint?
    if let Some(checkpoint) = args.resume_checkpoint {
        println!("Resuming from checkpoint {}", checkpoint);
        training = training.checkpoint(checkpoint);
    }
```

In terms of burn versions I'm using the latest source (as of today) - i.e. main branch (v0.22.0-pre.1)

Am I missing something or is it possible I've stumbled into a bug?

Thank you for a great project.

**Comment by laggui** [comment]
Oh, yeah this must be a bug. I think the problem arises with this code:

https://github.com/tracel-ai/burn/blob/2384148ef60a13fa09d2e1cfb56969d32e81b22b/crates/burn-train/src/learner/supervised/strategies/base.rs#L129-L138

which loads the checkpoint on the default device. The new `Device` holds autodiff as a property, so this might break somewhere else.
